BUG: Raise Warning if non-existent table is selected with `hdu=` on single-table FITS files

dhomeier opened the pull request:
Fixes an issue identified by @simonkrughoff :

On reading from a FITS file with only one `TABLE`-type extension, the `hdu=` argument is silently ignored when requesting a different HDU. In contrast, if there are multiple extensions, the appropriate `ValueError` is raised on selecting a non-existent or non-table HDU; a warning is given when no `hdu=` kwarg has been given.

This PR raises a warning about reading the default (i.e. first and only) table instead, if a different HDU was requested.
The risks for data integrity if the FITS file does not have the expected no. of table extensions might justify instead raising a `ValueError` here as well, but at a higher risk of breaking existing code, so it might be better to set a deprecation timeline for such behaviour.
Errors and warnings are a bit unspecific since they give the same message whether the HDU does not exist at all or is of a conflicting type.

Please assign to 4.1 if you feel this needs more discussion.

bsipocz commented:
I would say this is better go into 4.0, as a bugfix than to keep the old behaviour around  for 2 more years in LTS.

@eteq @astrofrog?

pllim commented:
There's one PEP 8 warning
```
astropy/io/fits/connect.py:187:71: W291 trailing whitespace
```

And some failures that surprise me. I did not expect this change to cause warnings to be raised in so many places.

bsipocz commented:
The failures are expected, one warning is raised when `hdu` is none, and the other when `hdu != 1`, that also includes the None case.

dhomeier commented:
> The failures are expected, one warning is raised when `hdu` is none, and the other when `hdu != 1`, that also includes the None case.

Good catch; should have run the full `io` and `table` tests! But the first version was not identifying the correct HDU in cases the first table was not in the first extension anyway. Added more tests for that configuration.

saimn commented on astropy/io/fits/connect.py:
```diff
@@ -181,6 +181,11 @@ def read_table_fits(input, hdu=None, astropy_native=False, memmap=False,
+                warnings.warn(f"No table found in specified hdu={hdu},"
+                              " reading in first available "
+                              f"table (hdu={first(tables)}) instead!",
+                              AstropyUserWarning)
```
You need a more strict test because currently `hdu in tables` gives unexpected results (I think there is a ticket about this). `hdu=5` or `hdu=''` will not raise the warning.

dhomeier replied:
Is that an `OrderedDict` issue? `hdu=''` wouldn't have surprised me too much, but `5`?
I cannot reproduce this with any Python 3.6-3.8 on MacOS X, and adding `5` and `''` to the tests still works, but if it fails on some platforms... what's a safe method for testing then? Would not need an `OrderedDict` in fact, since the standard dictionary already preserves order on 3.6+.

pllim replied:
It is ordered by insertion. Is it possible that insertion order is random?

dhomeier replied:
It's ordered by `enumerate(input)`, but `input` here is a `HDUList`, so if that has disorder issues that could be a problem.

saimn replied:
Oh sorry I missed that `tables` here is an OrderedDict, not the HDUList... forgot what I said :/

pllim replied:
Hmm, maybe not a problem with the order then? But I don't know what the comment about  lazy loading means, maybe @saimn knows.

https://github.com/astropy/astropy/blob/e2d877ec06745979b844a8d82bc137204726ae72/astropy/io/fits/hdu/hdulist.py#L239-L249

saimn replied:
This is the case that should not raise a warning (with a string argument):
```
In [12]: Table.read('astropy/io/fits/tests/data/memtest.fits', hdu='AXAF_CCDM')                
WARNING: No table found in specified hdu=AXAF_CCDM, reading in first available table (hdu=1) instead! [astropy.io.fits.connect]
Out[12]: 
<Table length=1>
       TIME       TLM_FMT  MJF   MNF   QUALITY [69]  ... CXPNAMOD CXPNARNG CXPNBMOD CXPNBRNG
        s                                            ...                                    
     float64       uint8  int32 int32      bool      ...  bytes4   bytes3   bytes4   bytes3 
----------------- ------- ----- ----- -------------- ... -------- -------- -------- --------
80348638.04702236       2  6887     0 False .. False ...     COHO       ON     COHO       ON
```

saimn replied:
@pllim @dhomeier - Sorry for the confusion, there is no ordering issue (unless I misunderstood something), it was just because of me believing that `tables` was the `HDUList`.
@pllim - In the `__iter__` loop the goal is to avoid loading all the HDUs when only one is requested, and for this `len` must not be called.

pllim replied:
Ok, ok. Ignore me then. Sorry for the noise.

dhomeier replied:
> In [12]: Table.read('astropy/io/fits/tests/data/memtest.fits', hdu='AXAF_CCDM')                
> WARNING: No table found in specified hdu=AXAF_CCDM, reading in first available table (hdu=1) instead! [astropy.io.fits.connect]

Strange, that's exactly the case that should be covered e.g. by `test_read_with_hdu_1['first']`. Only with multiple tables, but I just checked that this passes, too:
```
    @pytest.mark.parametrize('hdu', [None, 1, 'first'])
    def test_read_from_single_hdu(self, hdu):
        with catch_warnings() as l:
            t = Table.read(self.hdus[1])
        assert len(l) == 0
        assert equal_data(t, self.data1)
```

dhomeier replied:
I stand corrected; the test above worked reading directly from one `HDU`, not a `HDUList` with a single valid item. I can reproduce the failure now; will have to specifically check each `hdu.name` indeed.

saimn commented on astropy/io/fits/tests/test_connect.py:
```diff
@@ -273,8 +273,13 @@ def setup_class(self):
+        self.hdub = HDUList([hdu1, hdu4, hdu3, hdu2])
+        self.hdu3 = HDUList([hdu1, hdu4, hdu3])
+        self.hdu2 = HDUList([hdu1, hdu2, hdu4])
+        self.hdu1 = HDUList([hdu1, hdu2])
```
Using `self.hdu1` (singular, same for 2 and 3) is confusing. Could you change to `self.hdus1` (plural) ?

saimn commented on astropy/io/fits/tests/test_connect.py:
```diff
@@ -273,8 +273,13 @@ def setup_class(self):
         hdu1 = PrimaryHDU()
         hdu2 = BinTableHDU(self.data1, name='first')
         hdu3 = BinTableHDU(self.data2, name='second')
+        hdu4 = ImageHDU(np.ones((3, 3)), name='third')
```
It would be useful to test with a BinTableHDU without name.

saimn commented:
For me this is a bug, it should never read a different table than the one that was asked. So I would be happy with removing this behavior as a bugfix. 
Having a warning is fine too, but in this case I think that it should be a DeprecationWarning so we can remove this in the future.

astrofrog commented:
I would also treat this as a bug, and I think we could remove deprecate this behavior and remove in future.

dhomeier commented:
> I would also treat this as a bug, and I think we could remove deprecate this behavior and remove in future.

Shall I (just) change the type to `AstropyDeprecationWarning` or add a note about this becoming an error (in 4.1?)?

bsipocz commented:
Change the type please. Many  downstream are raising deprecation warning as errors, so they notice issues  early on, while they ignore UserWarnings altogether.

astrofrog commented:
@dhomeier - yes indeed change the type, and also make sure the message makes it clear that this is deprecated behavior.

dhomeier commented:
Hope I've addressed all testing requests; I've also modified the warning message according to non-table or entirely missing HDUs. Might also make sense to have similarly specific error messages, but this has already added some complexity to the tests (at least for me trying to sort them out between tricks and treats...)

pllim commented on CHANGES.rst:
```diff
@@ -661,6 +661,9 @@ astropy.io.fits
 - Implemented skip (after warning) of header cards with reserved
   keywords in ``table_to_hdu``. [#9390]
 
+- Add warning to ``read_table_fits`` when ``hdu=`` is selected, but
```
```suggestion
- Add ``AstropyDeprecationWarning`` to ``read_table_fits`` when ``hdu=`` is selected, but
```

dhomeier commented:
@saimn, all the tests requests should be addressed, if perhaps not very transparently...
I noted one case that is still not covered in the HDU check - if an `HDU` object is passed to `hdu=` and does not exist in the file/HDUList, a `ValueError` is raised instead of a `KeyError`. Seems a quite unlikely constellation...

saimn commented:
Yes, looks all good now, thanks @dhomeier :+1: 
And yes I agree about passing an HDU object to `hdu`, there is no reason to use this.
